### PR TITLE
Get rid of regexp and handle cursor in trim in doc printer

### PR DESCRIPTION
**Pull request by thorn0**
## Description

see https://github.com/prettier/prettier/pull/13415#issuecomment-1236294897

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/document/printer.js [hunk 1]
```diff
@@ -142,27 +142,41 @@ function generateInd(ind, newPart, options) {
...
-    const trimmed = last.replace(/[\t ]*$/, "");
-    trimCount += last.length - trimmed.length;
+    if (process.env.NODE_ENV !== "production" && typeof last !== "string") {
+      throw new Error(`Unexpected value in trim: '${typeof last}'`);
+    }
 
-    if (trimmed !== "") {
-      out[out.length - 1] = trimmed;
-      break;
+    for (let charIndex = last.length - 1; charIndex >= 0; charIndex--) {
```
Comment (fisker) [inline src/document/printer.js:163]: How do you think add a function to "get trailing space/tab length" so we don't need use label to break.

File: src/document/printer.js [hunk 2]
```diff
@@ -142,27 +142,41 @@ function generateInd(ind, newPart, options) {
...
 
-  while (out.length > 0) {
-    const last = getLast(out);
-    if (typeof last !== "string") {
-      break;
+  outer: while (outIndex--) {
+    const last = out[outIndex];
+
+    if (last === CURSOR_PLACEHOLDER) {
+      cursorCount++;
```
Comment (fisker) [inline src/document/printer.js:155]: I'd use an array to save cursors, and push them directly after text trimmed. But use number is fine too.

File: src/document/printer.js (hunk 1 above)
Comment (thorn0) [reply src/document/printer.js:163]: If avoiding the label is the only benefit, I'm not convinced.

**Review by fisker — APPROVED** [review]
Can you add a comment explain why we don't use `RegExp`?

**Pull request merged**